### DISCUSSION TITLE
Fixed: Compile of Tasmota 14.x.x fails for ESP32-Cx

### DISCUSSION THREAD

**Discussion by helmar74** [arendst/Tasmota#21964] [General]
Hi, I tried to compile Tasmota 14.2.0 for ESP32C3.

I only changed that in platformio_override.ini:
```
                          tasmota32c2
```

everything else is on "default"

I got the following error:
```
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libhelix-mp3\subband.c.o
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libhelix-mp3\trigtabs.c.o
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libmad\bit.c.o
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/users/helmar/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/12.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libmad\decoder.c.o
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libmad\fixed.c.o
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libmad\frame.c.o
Compiling .pio\build\tasmota32c3\lib89d\ESP8266Audio\libmad\huffman.c.o
*** [.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
========================================================================================================================== [FAILED] Took 133.65 seconds ==========================================================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32c3    FAILED    00:02:13.647
====================================================================================================================== 1 failed, 0 succeeded in 00:02:13.647 ======================================================================================================================

 *  Der Terminalprozess "C:\Users\helmar\.platformio\penv\Scripts\platformio.exe 'run'" wurde mit folgendem Exitcode beendet: 1. 
 *  Das Terminal wird von Aufgaben wiederverwendet, drücken Sie zum Schließen eine beliebige Taste. 
```

It worked before with 14.1.0

Edit from maintainer: Fixed in development branch with Tasmota Platform 2025.01.31

**Comment by arendst** [comment]
Ah windows and as.exe again.

Welcome to the club.

**Reply by helmar74** [reply]
ok, happy to be in the club... 
what does it mean?

**Comment by barbudor** [comment]
What exactly did you change ?

You shouldn't have to change anything but only select the correct target in VisualCode, and when platformio has done its trick, click the build button

**Comment by helmar74** [comment]
that's what i did. Comment out all target, only tasmota32c3 is choosen.

**Reply by barbudor** [reply]
That's what I said
You shoulddn't have to change anything in any file
Just select the target in the UI:
![image](https://github.com/user-attachments/assets/1780c08d-3878-4867-8e11-fb4ed8307cad)

**Comment by helmar74** [comment]
Will try:
[platformio]
core_dir = c:\.platformio
workspace_dir = c:\.pio

as mentioned here:
[https://github.com/arendst/Tasmota/discussions/21941](https://github.com/arendst/Tasmota/discussions/21941)
I'll report back

**Reply by barbudor** [reply]
Hum, I can reproduced
strange
there's no obvious reason

**Reply by barbudor** [reply]
it ends differently in my case but still related to as.exe

```
Compiling .pio\build\tasmota32c3\lib168\WiFi\WiFiAP.cpp.o
Archiving .pio\build\tasmota32c3\lib254\libNetwork.a
Compiling .pio\build\tasmota32c3\lib168\WiFi\WiFiGeneric.cpp.o
Archiving .pio\build\tasmota32c3\lib0f4\libUpdate.a
Compiling .pio\build\tasmota32c3\lib168\WiFi\WiFiMulti.cpp.o
Archiving .pio\build\tasmota32c3\libc26\libHTTPClient.a
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/users/jeanm/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/12.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
==================================================================== [FAILED] Took 75.68 seconds ====================================================================
Environment    Status    Duration
-------------  --------  ------------
tasmota32c3    FAILED    00:01:15.678
=============================================================== 1 failed, 0 succeeded in 00:01:15.678 ===============================================================
 *  The terminal process "C:\Users\jeanm\.platformio\penv\Scripts\platformio.exe 'run', '--environment', 'tasmota32c3'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. 
```

**Comment by arendst** [comment]
Try this https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry

**Reply by barbudor** [reply]
@arendst it's already there

**Reply by barbudor** [reply]
And I don't see any long path here. The path to as.exe, even with all ../.. is around 136 characters

**Comment by helmar74** [comment]
Changing
[platformio]
core_dir = c:.platformio
workspace_dir = c:.pio

in platformio_override fixed it.

Thanks a lot

**Comment by helmar74** [comment]
The error is back!
I tried to compile (only for test reasons) pre-release 14.3.0.
Im using VSC on Windows
I changed in registry to enable long paths:
<img width="869" alt="image" src="https://github.com/user-attachments/assets/21edff99-1ec3-4d60-9c38-57bbbfa5928e">


I also changed default directories in platform_override.ini:
```
; For best Gitpod performance remove the ";" in the next line. Needed Platformio files are cached and installed at first run
core_dir = c:\.platformio
; For unrelated compile errors with Windows it can help to shorten Platformio project path
workspace_dir = c:\.pio
```
When I try to compilfe for esp32c3 I get the following error:
```
.
.
.

Compiling C:\.pio\build\tasmota32c3\lib533\WiFi\WiFiMulti.cpp.o
Archiving C:\.pio\build\tasmota32c3\lib14d\libHTTPClient.a
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [C:\.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
========================================================================================================================== [FAILED] Took 178.49 seconds ==========================================================================================================================

Environment    Status    Duration
-------------  --------  ------------
```

It worked with Tasmota 14.2.0, but now the error seems to be back.

Compiling for ESP32 works without an issue

**Comment by helmar74** [comment]
see comment above

**Reply by helmar74** [reply]
Error also occurs with current development version 14.2.0.6:
```
Archiving .pio\build\tasmota32c3\lib240\libNetwork.a
riscv32-esp-elf-g++: fatal error: cannot execute 'C:/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
========================================================================================================================== [FAILED] Took 164.83 seconds ==========================================================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32c3    FAILED    00:02:44.826
===================================================================================================================== 1 failed, 0 succeeded in 00:02:44.826 ===================================================================================================================== 

```

**Comment by arendst** [comment]
It's not a Tasmota error. It's a platformio error.

In the end I moved to WSL on windows and run platformio on Ubuntu while running VSC in windows. It even compiles faster than pure windows.

**Reply by helmar74** [reply]
Hi and thanks for the info.
Sad to hear, as I can compile ESP8266, ESP32 and also tasmota32c2 under windows. I was also able to compile until version 14.2.0 on windows environment. Maybe it's a Core3 Issue for ESP32C3?

I only run 3 ESP32C3 based devices. So this devices will stay on 14.2.0.

**Comment by helmar74** [comment]
Is there anyone who is able to compile 14.2.0.6 on windows?

**Reply by Noschvie** [reply]
Here I am... Windows 10 Pro
VSC Info page:
```
Version: 1.94.2 (system setup)
Commit: 384ff7382de624fb94dbaf6da11977bba1ecd427
Datum: 2024-10-09T16:08:44.566Z
Electron: 30.5.1
ElectronBuildId: 10262041
Chromium: 124.0.6367.243
Node.js: 20.16.0
V8: 12.4.254.20-electron.0
Betriebssystem: Windows_NT x64 10.0.19045
```

**Reply by helmar74** [reply]
OK, so it looks like it should work also under windows...
Did you compile tasmota32c3 and did you do some modifications in platform_override.ini?

```
Version: 1.94.2 (system setup)
Commit: 384ff7382de624fb94dbaf6da11977bba1ecd427
Datum: 2024-10-09T16:08:44.566Z
Electron: 30.5.1
ElectronBuildId: 10262041
Chromium: 124.0.6367.243
Node.js: 20.16.0
V8: 12.4.254.20-electron.0
Betriebssystem: Windows_NT x64 10.0.26100
```

**Reply by ulfh2018** [reply]
I am running windows 11 pro latest update and VScode + Platformio. Downloaded latest version of Tasmota-development. I can compile ESP32, ESP32s2 and ESP32s3 without problems, but ESP32c3 fails. I am considering installing Ubuntu on my laptop and see if that is a solution, but my desktop PC is much better, and I would prefer to use it.

I get the followin error message using env:tasmota32c3:
riscv32-esp-elf-g++: fatal error: cannot execute 'C:/Users/ulfho/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1

**Reply by arendst** [reply]
Instead of installing ubuntu learn to use WSL. In that case you still run VSC on windows but platformio on WSL (which is default ubuntu).

**Reply by helmar74** [reply]
Same in my side, all ist working but tasmota32c3 fails with same error.
IT was able to compile in virtual machine running Linux Mint.
Not nice to run 2 development Machines...

**Reply by helmar74** [reply]
> Instead of installing ubuntu learn to use WSL. In that case you still run VSC on windows but platformio on WSL (which is default ubuntu).

Im running Backups of my Windows machine with dism tool. This fails when wsl is installed. I still want to Unterstand what is causing the issue in newer Versions as 14.2.0...

**Reply by ulfh2018** [reply]
Just tried to change the settings in platformio_override.ini acording to the recomendation in this tread. Worked like a dream!!!!
Thanks a lot to helmar74!!!!

**Reply by helmar74** [reply]
> Just tried to change the settings in platformio_override.ini acording to the recomendation in this tread. Worked like a dream!!!! Thanks a lot to helmar74!!!

Which Tasmota Version die you compile?
IT worked for me with 14.2.0, all later Versions failed.
You can also try to compile a other Version and than Switch back to tasmota32c3.
This also failed for me

**Reply by ulfh2018** [reply]
I am running 14.2.0.6
Tried to compile for esp8266 - which worked OK and the switched back to ESP32c3
Worked OK - though I hae not tested the image yet

**Reply by helmar74** [reply]
OK for me it worked Not with esp32c3. 
I did a cleanup of platformio but also No success.
Maybe i`ll give it a try in a new Maschine...

**Comment by helmar74** [comment]
Again gave it a try tasmota32c3 fails constantly:

```
Compiling C:\.pio\build\tasmota32c3\libef6\ESP8266Audio\libopus\silk\NLSF_decode.c.o
Compiling C:\.pio\build\tasmota32c3\libef6\ESP8266Audio\libopus\silk\NLSF_del_dec_quant.c.o
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [C:\.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
========================================================================================================================== [FAILED] Took 137.51 seconds ==========================================================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32c3    FAILED    00:02:17.508
===================================================================================================================== 1 failed, 0 succeeded in 00:02:17.508 ===================================================================================================================== 

 *  Der Terminalprozess "C:\.platformio\penv\Scripts\platformio.exe 'run'" wurde mit folgendem Exitcode beendet: 1. 
 *  Das Terminal wird von Aufgaben wiederverwendet, drücken Sie zum Schließen eine beliebige Taste. 
```

**Comment by ulfh2018** [comment]
I started the process with removing VSCode, using Revo Uninstaller, then I deleted the .platformio directory, rebooted and installed from scratch. Had several versions of Tasmota on my system, and deleted everything before I downloaded the .zip file, unzipped and moved the tasmota-development (bottom directory) to where I wanted it to be. Did not work until I changed the platforio_override.ini.

 

Next step would  have been to install Ubuntu on a virtual machine.

 

Fra: Helmar ***@***.***> 
Sendt: 11 October 2024 17:05
Til: arendst/Tasmota ***@***.***>
Kopi: ulfh2018 ***@***.***>; Comment ***@***.***>
Emne: Re: [arendst/Tasmota] Compile of Tasmota 14.2.0 failed for ESP32C3. UPDATE: Error is back on14.2.0.6 and 14.3.0 pre-release (Discussion #21964)

 

Again gave it a try tasmota32c3 fails constantly:

Compiling C:\.pio\build\tasmota32c3\libef6\ESP8266Audio\libopus\silk\NLSF_decode.c.o
Compiling C:\.pio\build\tasmota32c3\libef6\ESP8266Audio\libopus\silk\NLSF_del_dec_quant.c.o
riscv32-esp-elf-g++: fatal error: cannot execute 'c:/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [C:\.pio\build\tasmota32c3\src\tasmota.ino.cpp.o] Error 1
========================================================================================================================== [FAILED] Took 137.51 seconds ==========================================================================================================================
 
Environment    Status    Duration
-------------  --------  ------------
tasmota32c3    FAILED    00:02:17.508
===================================================================================================================== 1 failed, 0 succeeded in 00:02:17.508 ===================================================================================================================== 
 
 *  Der Terminalprozess "C:\.platformio\penv\Scripts\platformio.exe 'run'" wurde mit folgendem Exitcode beendet: 1. 
 *  Das Terminal wird von Aufgaben wiederverwendet, drücken Sie zum Schließen eine beliebige Taste.

**Comment by Jason2866** [comment]
Windows hits its limits with path length limitation. Platformio is just not doing crude workarounds for Windows to get big Projects going. Don't waste your time trying to workaround for a limitation Microsoft is not willing to solve since many many years. There are issue free alternatives 

The shorten of the project path will not work very long. With every library or source file added to Tasmota the command line for the compiler, linker and assembler will get longer.

**Reply by helmar74** [reply]
I enabled Long path names (32768 letters max) in registry,isn't enough length?

**Reply by Jason2866** [reply]
Nope 32k will not be enough.

**Reply by helmar74** [reply]
Ok inmy Linux VM IT works, so fine for me. Will Close discussion in a few days

**Reply by Jason2866** [reply]
Probably it is faster too, since Windows Defender is not anymore horribly slow down the compile process.

**Reply by Wintespe** [reply]
I ran the command **getconf -a** in the console under Linux Mint. This resulted in the following values


```
@Mint:~$ getconf -a
LINK_MAX                     65000
_POSIX_LINK_MAX              65000
MAX_CANON                    255
_POSIX_MAX_CANON             255
MAX_INPUT                    255
_POSIX_MAX_INPUT             255
NAME_MAX                     255
_POSIX_NAME_MAX              255
PATH_MAX                     4096
_POSIX_PATH_MAX              4096
```

**The description of the values**

LINK_MAX
  Specifies the maximum number of links that this file can have.
MAX_CANON
  Specifies the maximum number of bytes in the workstation's canonical input queue (before line editing).
MAX_INPUT
  Specifies the space available in the workstation's input queue.
NAME_MAX
  Specifies the largest filename size.
PATH
  Specifies the standard PATH setting.
_CS_PATH
  Specifies the standard PATH setting.
PATH_MAX
  Specifies the maximum number of bytes in a path name.


The result surprised me:
**NAME_MAX 4096
PATH 4096**
If I interpret this correctly, there are restrictions on the path length under Linux.
Does this mean that compiling projects on Linux - like Tasmota - under PlatformIO can run into problems?

**Reply by sfromis** [reply]
Well, 4KB is indeed a pretty typical max path length on Linux, meaning that projects are expected to not exceed this high limit, about 100 lines. I'm not aware of projects getting near this ceiling.

Where people have trouble is on Windows, where path names longer than ~256 seems to not be supported 100% of the time, but I'm not doing much in Windows anyway.

**Reply by Wintespe** [reply]
Well, with the help of Mark Russinovich's ProcMon, I took a look at what PlatformIO actually does when Tasmota is compiled. The result is amazing. I haven't found any path lengths longer than 168 bytes in my environment. However, the command line that is sent to as.exe exceeds a length of 4096 bytes, wow. Maybe that's where the problems lie. However, these depend on the PlafformIO.

**Comment by tefod-zz** [comment]
Do have exactly the same issues with v14.3.0.7 and only with the esp32-c3.
`riscv32-esp-elf-g++: fatal error: cannot execute 'c:/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe': CreateProcess: No such file or directory
compilation terminated.
*** [C:\.pio\build\tasmota32-c3\src\tasmota.ino.cpp.o] Error 1
Compiling C:\.pio\build\tasmota32-c3\lib9b0\lvgl\draw\vg_lite\lv_draw_buf_vg_lite.c.o`

Tried all mentioned workarounds - no luck so far. 
btw - all other variants (esp8266, esp32 etc.) do build.

**Reply by helmar74** [reply]
In the meantime I switch over to linux for compiling tasmota, which works well

**Reply by tefod-zz** [reply]
Thought about that. Are u using a VM or WSL? Honestly, I like the portability of VSC, which works great for me everywhere 🤷‍♂️

**Reply by arendst** [reply]
I prefer WSL. Works great.

**Reply by helmar74** [reply]
I'm using Linux Mint in a VM with VMware Workstation Personal.
I installed VSC in the VM. 
So I can use the advantages of VSC combined with error free comipilation

**Reply by Jason2866** [reply]
Yep, the only future proof solution for Windows users to compile Tasmota is using WSL or Linux in a VM.
The reason is the long generated command from Platformio when calling `as.exe`
When this command has more than 4095 bytes it fails.

**Reply by tefod-zz** [reply]
> I'm using Linux Mint in a VM with VMware Workstation Personal. I installed VSC in the VM. So I can use the advantages of VSC combined with error free comipilation

Thx. For some specific projects, I even use Alpine Linux, so I'm not completely new to Linux 😁
However, portability is an issue for me since I can't install VMware on every machine, especially at work. Additionally, the overhead of a full Linux OS can be impractical. Even Alpine with Xfce would take around a gigabyte.

I haven't explored WSL yet, but it seems like I should give it a try now. In the meantime, I'm thinking of setting up an Alpine VM or appliance with VSC on my home lab (proxmox). That way, I could access it via RDP/SSH & SCP anytime.

**Comment by ulfh2018** [comment]
Just compiled tasmota32c3 on windows 11. Have placed the following entries in platformio.ini:

 

core_dir = C:.platformio

workspace_dir = C:.pio

 

This should be put under the section [platformio]

Fra: tefo ***@***.***> 
Sendt: 02 December 2024 12:41
Til: arendst/Tasmota ***@***.***>
Kopi: ulfh2018 ***@***.***>; Comment ***@***.***>
Emne: Re: [arendst/Tasmota] Compile of Tasmota 14.2.0 failed for ESP32C3. UPDATE: Error is back on14.2.0.6 and 14.3.0 pre-release (Discussion #21964)

 

I'm using Linux Mint in a VM with VMware Workstation Personal. I installed VSC in the VM. So I can use the advantages of VSC combined with error free comipilation
Thx. For some specific projects, I even use Alpine Linux, so I'm not completely new to Linux 😁
However, portability is an issue for me since I can't install VMware on every machine, especially at work. Additionally, the overhead of a full Linux OS can be impractical. Even Alpine with Xfce would take around a gigabyte.

I haven't explored WSL yet, but it seems like I should give it a try now. In the meantime, I'm thinking of setting up an Alpine VM or appliance with VSC on my home lab (proxmox). That way, I could access it via RDP/SSH & SCP anytime.

**Reply by helmar74** [reply]
This worked for me also until tasmota 14.2.0. From 14.3.0 on it wasn't working anymore.
Just want to mention, that compiling with linux (in a VM) is much faster as compiling on a physical windows machine.
That sound's cracy, but is a fact ;-)

**Comment by joba-1** [comment]
you won‘t notice much differeces between using a vm hosting environment like vmware or wsl once you have setup the vm properly. 
But the integration of wsl is better imo since distros provide special images taylored for it. You usually get networking setup, can run X Windows gui programs seamlessly on the windows desktop, have a share of windows drive in wsl and of the wsl drive in windows.
Opening a wsl terminal does not require an additional program like putty. All just by installing the distro via a simple wsl command. No further setup needed to get all that.

Regarding VSCode: you have the option of running the linux variant in WSL or use the remote dev extension with the windows variant. I prefer the latter.

Have fun…

**Comment by Wintespe** [comment]
As I noted above, the problem is probably the length of the command line that PlatformIO sends to as. It exceeds 4095 bytes. The limit for Windows is 8k. As far as I know, the limit for Linux is 32k. The other ID's, such as VisualStudio 2022 etc, work around the problem by creating a file with the commands and passing it to the appropriate tool . I hope Platformio will add this functionality in one of the next versions. At some point, even at 32k, the limit will be exceeded with the command line. For now I'll wait until it burns out on me.

**Comment by Jason2866** [comment]
Platformio development seems stalled (it is an Ukraine company)

**Reply by Wintespe** [reply]
Yes, I know that. Hopefully the disaster will stop.

**Comment by juepi** [comment]
Hi Folks,

I want to mention a relatively simple workaround for the windows path issue that helped me to get rid of the "as.exe file not found issue":

a.) Issue the following command in Windows terminal (not elevated, same user you run VSC, adopt to your needs):
`subst p: C:\Users\Your_Username\.platformio`

b.) add the following to your `[platformio]` section of `platformio.ini`:
`core_dir = p:\`

--> no more path issues for me, firmware for ESP32-C6 compiles correctly now. 😉

Same thing could be done with the `workspace_dir` if required i assume.
Maybe this could be done automagically with some pre/post commands?

yours,
Juergen

**Reply by Wintespe** [reply]
High Juergen,

as far as I know, this is not a path length problem. The problem comes from the 8k command line limitation.
As a result of substituting the path into a drive, the command line to the assembler becomes significantly shorter. The command line for as.exe such as the following;

```
<ProcessName>as.exe</ProcessName>
<ImagePath>C:\Users\Win10\.platformio\packages\toolchain-riscv32-esp\riscv32-esp-elf\bin\as.exe</ImagePath>
<CommandLine>C:/Users/Win10/.platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/13.2.0/../../../../riscv32-esp-elf/bin/as.exe -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/libraries/Wire/src -I include -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32c3/include/newlib/platform_include -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32c3/include/freertos/config/include -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32c3/include/freertos/config/include/freertos -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32c3/include/freertos/config/riscv/include -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32c3/include/freertos/FreeRTOS-Kernel/include -I C:/Users/Win10/.platformio/packages/framework-arduinoespressif32/tools/esp32-arduino-libs/esp32c3/include/freertos/FreeRTOS-Kernel/portable/riscv/include .......
```
will be shorter.

The procedure helps temporarily but does not solve the problem permanently. PlatformIO probably needs to be adapted for this. It might also help if the feature set of the ESP32 versions of Tasmota could be reduced by simple means.

**Reply by Jason2866** [reply]
> It might also help if the feature set of the ESP32 versions of Tasmota could be reduced by simple means.

That's not gonna happen

**Reply by Wintespe** [reply]
I mean with the reduction of the features that a user-specific compilation for the ESP32 applications, as is possible with the ESP8266 versions by using user_config_override.h, would be very helpful. Some previous discussions show a problem there because of the increased familiarisation phase.

**Reply by juepi** [reply]
@Wintespe that's why i named my suggestion a _workaround_ 😉
I fully agree that a solution would address underlying issue and - probably - needs to be fixed by the PIO team, but as a workaround (which at least temporarily solved my problem), i find substituting a drive a little less messy.

**Reply by sfromis** [reply]
I suspect that it is not a high priority to deal with the restrictions imposed by the limitations of Windows, as long as it is at least *possible* to do a build with workarounds....

**Reply by tefod-zz** [reply]
I’ve spent a lot of time thinking about how to tackle this and similar 'issues' in a sustainable way while still being system- and device-independent. The goal is to work on the same code, kind of portable and from different locations. This doesn’t just apply to Tasmota but also to other projects like WLED, ESP32-CAM, etc.

I’ve found two possible solutions but haven’t decided yet which one to use:
a) [code-server](https://github.com/coder/code-server): _Run VS Code on any machine and access it through a browser._
b) [remote-SSH](https://code.visualstudio.com/docs/remote/ssh)

**Comment by Jason2866** [comment]
@tefod-zz Both ways you mentioned have a no go limitation. Flashing the esp is not possible

Edit: The only real solution has to be done from Platformio. Shorten the path by using an env variable or storing the needed commands in a file and call the "file" instead of a looong command line.
https://stackoverflow.com/questions/7139685/does-gcc-support-command-files

**Reply by tefod-zz** [reply]
I'm not sure if we can really influence the situation with PlatformIO, as you mentioned earlier that development seems to be stalled at the moment.  The limitation of not being able to flash directly could be considered manageable, as flashing can easily with external tools or via OTA 🤷‍♂️

**Reply by Jason2866** [reply]
No, flashing via VSC is one of the big advantages. In other words no alternative will be even considered not having this function easy available.
There **is** already a 100% working solution for this issue. Installing and using WSL.

**Reply by tefod-zz** [reply]
It seems to depend on individual requirements. For me, flashing with VSC is not crucial at all. I have over 20 ESP32 devices installed at home; using VSC for flashing wouldn’t even be feasible. I handle it via OTA, or when necessary, with [tools ](https://tasmota.github.io/docs/Getting-Started/#needed-software) like Tasmotizer. VSC with WSL is a workaround but comes with a [3GB overhead](https://superuser.com/a/1712687/1151303) and, as far as I know, [doesn’t work alongside](https://superuser.com/a/1829372/1151303) VMware.

**Reply by Jason2866** [reply]
As always, it depends. Still talking as maintainer of the Platform and framework, the inbuilt flash function has to work.

**Comment by Jason2866** [comment]
@helmar74 @Wintespe @tefod-zz and all other who wants to test a possible Windows path length Platformio fix.

Change in `platformio_override.ini` the entry 
```
[env:tasmota32_base]
; *** Uncomment next lines ";" to enable development Tasmota Arduino version ESP32
;platform                = https://github.com/Jason2866/platform-espressif32/releases/download/2024.07.22/platform-espressif32.zip
```
to
```
[env:tasmota32_base]
; *** Uncomment next lines ";" to enable development Tasmota Arduino version ESP32
platform                = https://github.com/tasmota/platform-espressif32.git#Arduino/IDF53
```
Please post if the issue is solved with this fix.

**Reply by arendst** [reply]
Yep. Solves it for me.

**Reply by Wintespe** [reply]
I get the error message that I should install a **Git client**, which does not help, a pity.

Processing tasmota32c3 (board: esp32c3; framework: arduino; platform: https://github.com/tasmota/platform-espressif32.git#Arduino/IDF53)
-------------------------------------------------------------------------------------------------------------------------------------------------------------Platform Manager: Installing git+https://github.com/tasmota/platform-espressif32.git#Arduino/IDF53
**UserSideException: Please install Git client from https://git-scm.com/downloads**

 *  The terminal process "C:\Users\Win10\.platformio\penv\Scripts\platformio.exe 'run', '--environment', 'tasmota32c3'" terminated with exit code: 1.

**Reply by helmar74** [reply]
> @helmar74 @Wintespe @tefod-zz and all other who wants to test a possible Windows path length Platformio fix.
> 
> Change in `platformio_override.ini` the entry
> 
> ```
> [env:tasmota32_base]
> ; *** Uncomment next lines ";" to enable development Tasmota Arduino version ESP32
> ;platform                = https://github.com/Jason2866/platform-espressif32/releases/download/2024.07.22/platform-espressif32.zip
> ```
> 
> to
> 
> ```
> [env:tasmota32_base]
> ; *** Uncomment next lines ";" to enable development Tasmota Arduino version ESP32
> platform                = https://github.com/tasmota/platform-espressif32.git#Arduino/IDF53
> ```
> 
> Please post if the issue is solved with this fix.

Doesn't work for me.
I have downloaded current tasmota development.
Changed the following in platform_override.ini:
<img width="556" alt="image" src="https://github.com/user-attachments/assets/560d54ff-1819-432b-8dbb-3f5fe8e8b25d" />

and 

<img width="194" alt="image" src="https://github.com/user-attachments/assets/d5819399-d4ee-41db-8ac1-667c47833ab6" />

ends up in error:

<img width="538" alt="image" src="https://github.com/user-attachments/assets/ac3639fd-349f-4f43-a566-36485f325d42" />

**Reply by Jason2866** [reply]
@helmar74 Close VSC. Delete project cache folder. Delete `.platformio/platforms` and `platformio/packages`
Start VSC and wait until finished. Compile.

**Reply by Jason2866** [reply]
@Wintespe git client not correctly installed. Probably not in path. 
You can do without git by following
```
platform = https://github.com/tasmota/platform-espressif32/archive/refs/heads/Arduino/IDF53.zip
```

**Reply by helmar74** [reply]
> @helmar74 Close VSC. Delete project cache folder. Delete `.platformio/platforms` and `platformio/packages` Start VSC and wait until finished. Compile.

using this worked for me (compiling tasmota32c3):
<img width="523" alt="image" src="https://github.com/user-attachments/assets/92959972-1b68-4d37-8e6f-0779f791b7c5" />

deleting the folders project cache folder. `.platformio/platforms` and `platformio/packages` doesn't work for me

**Reply by Wintespe** [reply]
@Jason2866
I have and will probably not install Git.
Thanks for the additional link:
https://github.com/tasmota/platform-espressif32/archive/refs/heads/Arduino/IDF53.zip
I was able to compile tasmota32solo1 without errors, great. Whether the binary works I will test the days.
What was the trick that it now works under Windows?

**Reply by Jason2866** [reply]
@Wintespe https://github.com/tasmota/platform-espressif32/commit/cadff826d4f019ee7178bd9562fd1ecd54b1c99e

**Reply by tefod-zz** [reply]
> Please post if the issue is solved with this fix.

Thx mate - it works flawlessly, and I can now compile Tasmota32-C3 on Windows without any issues. This will save me a lot of time for using workarounds.

Since I couldn't find a buymeacoffee.com link from you, I decided to honor your time by making a small donation to the German Children's Cancer Aid instead.
Much appreciated for your support and contributions!

![image](https://github.com/user-attachments/assets/f8b39065-ac7f-41e0-8ae2-7e0b67c2cb1e)

**Reply by Jason2866** [reply]
Thank you very much for the donation.

**Comment by Jason2866** [comment]
Fixed in branch development with #22832

**Reply by barbudor** [reply]
Thanks @Jason2866  🙇‍♂️
